runtime/ops/io.rs

```diff
@@ -142,6 +156,80 @@ pub struct FileMetadata {
   pub tty: TTYMetadata,
 }
 
+#[derive(Debug)]
+pub struct WriteOnlyResource<S> {
+  stream: AsyncRefCell<S>,
+}
+
+impl<S: 'static> From<S> for WriteOnlyResource<S> {
+  fn from(stream: S) -> Self {
+    Self {
+      stream: stream.into(),
+    }
+  }
+}
+
+impl<S> WriteOnlyResource<S>
+where
+  S: AsyncWrite + Unpin + 'static,
+{
+  pub fn borrow_mut(self: &Rc<Self>) -> AsyncMutFuture<S> {
+    RcRef::map(self, |r| &r.stream).borrow_mut()
+  }
+
+  async fn write(self: &Rc<Self>, buf: &[u8]) -> Result<usize, AnyError> {
+    let mut stream = self.borrow_mut().await;
+    let nwritten = stream.write(buf).await?;
+    Ok(nwritten)
+  }
+
+  async fn shutdown(self: &Rc<Self>) -> Result<(), AnyError> {
+    let mut stream = self.borrow_mut().await;
+    stream.shutdown().await?;
+    Ok(())
+  }
+}
+
+#[derive(Debug)]
+pub struct ReadOnlyResource<S> {
+  stream: AsyncRefCell<S>,
+  cancel_handle: CancelHandle,
+}
+
+impl<S: 'static> From<S> for ReadOnlyResource<S> {
```

Comment:
These two From<S> -> ReadOnlyResource<S> seem backwards to me.... I would expect you to be able to go from


ReadOnlyResource<S> -> S

But not from any static lifetime object to ReadOnlyResource... am I missing something?

Reply:
I can add some type constraints (`S: AsyncRead + Unpin + 'static`), but these are not necessary.
Note that these type constraints are present on the `impl<S> ReadOnlyResource<S>` block.